File: decompiled/sources/com/hjq/shape/drawable/ShapeState.java
package com.hjq.shape.drawable;

import android.content.res.Resources;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;

public class ShapeState extends Drawable.ConstantState {
    public float mCenterX = 0.5f;
    public float mCenterY = 0.5f;
    public int mChangingConfigurations;
    public int[] mGradientColors;
    public ShapeGradientOrientation mGradientOrientation = ShapeGradientOrientation.TOP_BOTTOM;
    public float mGradientRadius = 0.5f;
    public int mGradientType = 0;
    public boolean mHasSolidColor;
    public int mHeight = -1;
    public int mInnerRadius;
    public float mInnerRadiusRatio;
    public boolean mOpaque;
    public Rect mPadding;
    public float[] mPositions;
    public float mRadius;
    public float[] mRadiusArray;
    public int mShadowColor;
    public int mShadowOffsetX;
    public int mShadowOffsetY;
    public int mShadowSize;
    public int mShapeType = 0;
    public int mSolidColor;
    public int mStrokeColor;
    public float mStrokeDashGap;
    public float mStrokeDashWidth;
    public int mStrokeWidth = -1;
    public int[] mTempColors;
    public float[] mTempPositions;
    public int mThickness;
    public float mThicknessRatio;
    public boolean mUseLevel;
    public boolean mUseLevelForShape;
    public int mWidth = -1;

    private static boolean isOpaque(int i) {
        return ((i >> 24) & 255) == 255;
    }

    public ShapeState() {
    }

    public ShapeState(ShapeState shapeState) {
        this.mChangingConfigurations = shapeState.mChangingConfigurations;
        this.mShapeType = shapeState.mShapeType;
        this.mGradientType = shapeState.mGradientType;
        this.mGradientOrientation = shapeState.mGradientOrientation;
        int[] iArr = shapeState.mGradientColors;
        if (iArr != null) {
            this.mGradientColors = (int[]) iArr.clone();
        }
        float[] fArr = shapeState.mPositions;
        if (fArr != null) {
            this.mPositions = (float[]) fArr.clone();
        }
        this.mHasSolidColor = shapeState.mHasSolidColor;
        this.mSolidColor = shapeState.mSolidColor;
        this.mStrokeWidth = shapeState.mStrokeWidth;
        this.mStrokeColor = shapeState.mStrokeColor;
        this.mStrokeDashWidth = shapeState.mStrokeDashWidth;
        this.mStrokeDashGap = shapeState.mStrokeDashGap;
        this.mRadius = shapeState.mRadius;
        float[] fArr2 = shapeState.mRadiusArray;
        if (fArr2 != null) {
            this.mRadiusArray = (float[]) fArr2.clone();
        }
        if (shapeState.mPadding != null) {
            this.mPadding = new Rect(shapeState.mPadding);
        }
        this.mWidth = shapeState.mWidth;
        this.mHeight = shapeState.mHeight;
        this.mInnerRadiusRatio = shapeState.mInnerRadiusRatio;
        this.mThicknessRatio = shapeState.mThicknessRatio;
        this.mInnerRadius = shapeState.mInnerRadius;
        this.mThickness = shapeState.mThickness;
        this.mCenterX = shapeState.mCenterX;
        this.mCenterY = shapeState.mCenterY;
        this.mGradientRadius = shapeState.mGradientRadius;
        this.mUseLevel = shapeState.mUseLevel;
        this.mUseLevelForShape = shapeState.mUseLevelForShape;
        this.mOpaque = shapeState.mOpaque;
        this.mShadowSize = shapeState.mShadowSize;
        this.mShadowColor = shapeState.mShadowColor;
        this.mShadowOffsetX = shapeState.mShadowOffsetX;
        this.mShadowOffsetY = shapeState.mShadowOffsetY;
    }

    public Drawable newDrawable() {
        return new ShapeDrawable(this);
    }

    public Drawable newDrawable(Resources resources) {
        return new ShapeDrawable(this);
    }

    public int getChangingConfigurations() {
        return this.mChangingConfigurations;
    }

    public void setShape(int i) {
        this.mShapeType = i;
        computeOpacity();
    }

    public void setGradientType(int i) {
        this.mGradientType = i;
    }

    public void setGradientCenter(float f, float f2) {
        this.mCenterX = f;
        this.mCenterY = f2;
    }

    public void setGradientColor(int[] iArr) {
        this.mHasSolidColor = false;
        this.mGradientColors = iArr;
        computeOpacity();
    }

    public void setSolidColor(int i) {
        this.mHasSolidColor = true;
        this.mSolidColor = i;
        this.mGradientColors = null;
        computeOpacity();
    }

    private void computeOpacity() {
        if (this.mShapeType != 0) {
            this.mOpaque = false;
        } else if (this.mRadius > 0.0f || this.mRadiusArray != null) {
            this.mOpaque = false;
        } else if (this.mStrokeWidth > 0 && !isOpaque(this.mStrokeColor)) {
            this.mOpaque = false;
        } else if (this.mHasSolidColor) {
            this.mOpaque = isOpaque(this.mSolidColor);
        } else {
            int[] iArr = this.mGradientColors;
            if (iArr != null) {
                for (int isOpaque : iArr) {
                    if (!isOpaque(isOpaque)) {
                        this.mOpaque = false;
                        return;
                    }
                }
            }
            this.mOpaque = true;
        }
    }

    public void setStroke(int i, int i2) {
        this.mStrokeWidth = i;
        this.mStrokeColor = i2;
        computeOpacity();
    }

    public void setStroke(int i, int i2, float f, float f2) {
        this.mStrokeWidth = i;
        this.mStrokeColor = i2;
        this.mStrokeDashWidth = f;
        this.mStrokeDashGap = f2;
        computeOpacity();
    }

    public void setCornerRadius(float f) {
        if (f < 0.0f) {
            f = 0.0f;
        }
        this.mRadius = f;
        this.mRadiusArray = null;
    }

    public void setCornerRadii(float[] fArr) {
        this.mRadiusArray = fArr;
        if (fArr == null) {
            this.mRadius = 0.0f;
        }
    }

    public void setSize(int i, int i2) {
        this.mWidth = i;
        this.mHeight = i2;
    }

    public void setGradientRadius(float f) {
        this.mGradientRadius = f;
    }

    public void setShadowColor(int i) {
        this.mShadowColor = i;
    }

    public void setShadowSize(int i) {
        this.mShadowSize = i;
    }

    public void setShadowOffsetX(int i) {
        this.mShadowOffsetX = i;
    }

    public void setShadowOffsetY(int i) {
        this.mShadowOffsetY = i;
    }
}

File: decompiled/sources/com/hjq/shape/drawable/ShapeState.kt
package com.hjq.shape.drawable

import android.content.res.Resources
import android.graphics.Rect
import android.graphics.drawable.Drawable

class ShapeState() : Drawable.ConstantState() {
    @JvmField
    var mCenterX = 0.5f

    @JvmField
    var mCenterY = 0.5f

    @JvmField
    var mChangingConfigurations = 0

    @JvmField
    var mGradientColors: IntArray? = null

    @JvmField
    var mGradientOrientation: ShapeGradientOrientation = ShapeGradientOrientation.TOP_BOTTOM

    @JvmField
    var mGradientRadius = 0.5f

    @JvmField
    var mGradientType = 0

    @JvmField
    var mHasSolidColor = false

    @JvmField
    var mHeight = -1

    @JvmField
    var mInnerRadius = 0

    @JvmField
    var mInnerRadiusRatio = 0f

    @JvmField
    var mOpaque = false

    @JvmField
    var mPadding: Rect? = null

    @JvmField
    var mPositions: FloatArray? = null

    @JvmField
    var mRadius = 0f

    @JvmField
    var mRadiusArray: FloatArray? = null

    @JvmField
    var mShadowColor = 0

    @JvmField
    var mShadowOffsetX = 0

    @JvmField
    var mShadowOffsetY = 0

    @JvmField
    var mShadowSize = 0

    @JvmField
    var mShapeType = 0

    @JvmField
    var mSolidColor = 0

    @JvmField
    var mStrokeColor = 0

    @JvmField
    var mStrokeDashGap = 0f

    @JvmField
    var mStrokeDashWidth = 0f

    @JvmField
    var mStrokeWidth = -1

    @JvmField
    var mTempColors: IntArray? = null

    @JvmField
    var mTempPositions: FloatArray? = null

    @JvmField
    var mThickness = 0

    @JvmField
    var mThicknessRatio = 0f

    @JvmField
    var mUseLevel = false

    @JvmField
    var mUseLevelForShape = false

    @JvmField
    var mWidth = -1

    constructor(shapeState: ShapeState) : this() {
        mChangingConfigurations = shapeState.mChangingConfigurations
        mShapeType = shapeState.mShapeType
        mGradientType = shapeState.mGradientType
        mGradientOrientation = shapeState.mGradientOrientation
        shapeState.mGradientColors?.let {
            mGradientColors = it.clone()
        }
        shapeState.mPositions?.let {
            mPositions = it.clone()
        }
        mHasSolidColor = shapeState.mHasSolidColor
        mSolidColor = shapeState.mSolidColor
        mStrokeWidth = shapeState.mStrokeWidth
        mStrokeColor = shapeState.mStrokeColor
        mStrokeDashWidth = shapeState.mStrokeDashWidth
        mStrokeDashGap = shapeState.mStrokeDashGap
        mRadius = shapeState.mRadius
        shapeState.mRadiusArray?.let {
            mRadiusArray = it.clone()
        }
        shapeState.mPadding?.let {
            mPadding = Rect(it)
        }
        mWidth = shapeState.mWidth
        mHeight = shapeState.mHeight
        mInnerRadiusRatio = shapeState.mInnerRadiusRatio
        mThicknessRatio = shapeState.mThicknessRatio
        mInnerRadius = shapeState.mInnerRadius
        mThickness = shapeState.mThickness
        mCenterX = shapeState.mCenterX
        mCenterY = shapeState.mCenterY
        mGradientRadius = shapeState.mGradientRadius
        mUseLevel = shapeState.mUseLevel
        mUseLevelForShape = shapeState.mUseLevelForShape
        mOpaque = shapeState.mOpaque
        mShadowSize = shapeState.mShadowSize
        mShadowColor = shapeState.mShadowColor
        mShadowOffsetX = shapeState.mShadowOffsetX
        mShadowOffsetY = shapeState.mShadowOffsetY
    }

    override fun newDrawable(): Drawable {
        return ShapeDrawable(this)
    }

    override fun newDrawable(res: Resources?): Drawable {
        return ShapeDrawable(this)
    }

    override fun getChangingConfigurations(): Int {
        return mChangingConfigurations
    }

    fun setShape(shape: Int) {
        mShapeType = shape
        computeOpacity()
    }

    fun setGradientType(gradientType: Int) {
        mGradientType = gradientType
    }

    fun setGradientCenter(centerX: Float, centerY: Float) {
        mCenterX = centerX
        mCenterY = centerY
    }

    fun setGradientColor(gradientColors: IntArray?) {
        mHasSolidColor = false
        mGradientColors = gradientColors
        computeOpacity()
    }

    fun setSolidColor(solidColor: Int) {
        mHasSolidColor = true
        mSolidColor = solidColor
        mGradientColors = null
        computeOpacity()
    }

    fun setStroke(strokeWidth: Int, strokeColor: Int) {
        mStrokeWidth = strokeWidth
        mStrokeColor = strokeColor
        computeOpacity()
    }

    fun setStroke(strokeWidth: Int, strokeColor: Int, strokeDashWidth: Float, strokeDashGap: Float) {
        mStrokeWidth = strokeWidth
        mStrokeColor = strokeColor
        mStrokeDashWidth = strokeDashWidth
        mStrokeDashGap = strokeDashGap
        computeOpacity()
    }

    fun setCornerRadius(radius: Float) {
        var safeRadius = radius
        if (safeRadius < 0f) {
            safeRadius = 0f
        }
        mRadius = safeRadius
        mRadiusArray = null
    }

    fun setCornerRadii(radiusArray: FloatArray?) {
        mRadiusArray = radiusArray
        if (radiusArray == null) {
            mRadius = 0f
        }
    }

    fun setSize(width: Int, height: Int) {
        mWidth = width
        mHeight = height
    }

    fun setGradientRadius(gradientRadius: Float) {
        mGradientRadius = gradientRadius
    }

    fun setShadowColor(shadowColor: Int) {
        mShadowColor = shadowColor
    }

    fun setShadowSize(shadowSize: Int) {
        mShadowSize = shadowSize
    }

    fun setShadowOffsetX(shadowOffsetX: Int) {
        mShadowOffsetX = shadowOffsetX
    }

    fun setShadowOffsetY(shadowOffsetY: Int) {
        mShadowOffsetY = shadowOffsetY
    }

    private fun computeOpacity() {
        if (mShapeType != 0) {
            mOpaque = false
        } else if (mRadius > 0f || mRadiusArray != null) {
            mOpaque = false
        } else if (mStrokeWidth > 0 && !isOpaque(mStrokeColor)) {
            mOpaque = false
        } else if (mHasSolidColor) {
            mOpaque = isOpaque(mSolidColor)
        } else {
            val gradientColors = mGradientColors
            if (gradientColors != null) {
                for (color in gradientColors) {
                    if (!isOpaque(color)) {
                        mOpaque = false
                        return
                    }
                }
            }
            mOpaque = true
        }
    }

    companion object {
        private fun isOpaque(color: Int): Boolean {
            return (color shr 24 and 255) == 255
        }
    }
}
